preinit_connection.go

```diff
@@ -0,0 +1,245 @@
+	).Error("Failed during connection handshake.")
+
+	if ne, ok := err.(net.Error); ok && ne.Timeout() {
+		err = ErrTimeout
+	}
+	if err == io.EOF {
+		err = NewWrappedSystemError(ErrCodeNetwork, io.EOF)
+	}
+	ch.writeMessage(c, &errorMessage{
+		id:      id,
+		errCode: GetSystemErrorCode(err),
+		message: err.Error(),
+	})
+	c.Close()
+	return err
+}
+
+func (ch *Channel) writeMessage(c net.Conn, msg message) error {
+	frame := ch.connectionOptions.FramePool.Get()
+	defer ch.connectionOptions.FramePool.Release(frame)
+
+	if err := frame.write(msg); err != nil {
+		return err
+	}
+	return frame.WriteOut(c)
+}
+
+func (ch *Channel) readMessage(c net.Conn, msg message) (uint32, error) {
+	frame := ch.connectionOptions.FramePool.Get()
+	defer ch.connectionOptions.FramePool.Release(frame)
+
+	if err := frame.ReadIn(c); err != nil {
+		return 0, err
+	}
+
+	if frame.Header.messageType != msg.messageType() {
+		if frame.Header.messageType == messageTypeError {
+			return 0, readError(frame)
+		}
+		return 0, NewSystemError(ErrCodeProtocol, "expected message type %v, got %v", msg.messageType(), frame.Header.messageType)
+	}
+
+	return frame.Header.ID, frame.read(msg)
+}
+
+func parseRemotePeer(p initParams, remoteAddr net.Addr) (PeerInfo, peerAddressComponents, error) {
+	var (
+		remotePeer        PeerInfo
+		remotePeerAddress peerAddressComponents
+		ok                bool
+	)
+
+	if remotePeer.HostPort, ok = p[InitParamHostPort]; !ok {
+		return remotePeer, remotePeerAddress, fmt.Errorf("header %v is required", InitParamHostPort)
+	}
+	if remotePeer.ProcessName, ok = p[InitParamProcessName]; !ok {
+		return remotePeer, remotePeerAddress, fmt.Errorf("header %v is required", InitParamProcessName)
+	}
+
+	// If the remote host:port is ephemeral, use the socket address as the
+	// host:port and set IsEphemeral to true.
+	if isEphemeralHostPort(remotePeer.HostPort) {
+		remotePeer.HostPort = remoteAddr.String()
+		remotePeer.IsEphemeral = true
+	}
+
+	remotePeer.Version.Language = p[InitParamTChannelLanguage]
+	remotePeer.Version.LanguageVersion = p[InitParamTChannelLanguageVersion]
+	remotePeer.Version.TChannelVersion = p[InitParamTChannelVersion]
+
+	address := remotePeer.HostPort
+	if sHost, sPort, err := net.SplitHostPort(address); err == nil {
+		address = sHost
+		if p, err := strconv.ParseUint(sPort, 10, 16); err == nil {
+			remotePeerAddress.port = uint16(p)
+		}
+	}
+	if address == "localhost" {
+		remotePeerAddress.ipv4 = 127<<24 | 1
+	} else if ip := net.ParseIP(address); ip != nil {
+		if ip4 := ip.To4(); ip4 != nil {
+			remotePeerAddress.ipv4 = binary.BigEndian.Uint32(ip4)
+		} else {
+			remotePeerAddress.ipv6 = address
+		}
+	} else {
+		remotePeerAddress.hostname = address
+	}
+
+	return remotePeer, remotePeerAddress, nil
+}
+
+func setInitDeadline(ctx context.Context, c net.Conn) func() {
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		deadline = time.Now().Add(5 * time.Second)
+	}
+
+	c.SetDeadline(deadline)
```

Comment:
If we have a deadline in the context, we use that, otherwise we use 5 seconds as a default timeout for inits.

We always want to set a deadline on the connection though to avoid connection exchanges that never end